Add AC Dimmer support

OttoWinter opened the pull request:
Fixes https://github.com/esphome/feature-requests/issues/278

## Description:

Thought I'd take a stab at this - it's the most requested feature in the FR tracker! I was skeptical at first how clean this could be done code-wise, but ended up being not so bad. Even multiple dimmers are supported well with this code.

Thanks also to @glmnet for the idea to attach onto waveform timer1 interrupt. With a bit of a rethink of how the interrupt works (as opposed to the programming in the examples by RoboDyn), it worked out rather well.

Not tested yet - I have some triacs here but no spare mains cable I can cut open. Will test with a function generator at the local makerspace here later.

You can test this code by copying [this directory](https://github.com/esphome/esphome/tree/ac-dimmer/esphome/components/dimmer) into a `custom_components` folder in your ESPHome config folder (see also contributing guide).

Also not sure about the name - `dimmer` or `ac_dimmer`, comments welcome.


**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/536

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs/pull/536).

OttoWinter committed:
Add AC Dimmer support

Fixes https://github.com/esphome/feature-requests/issues/278

OttoWinter committed:
Add AC Dimmer support

Fixes https://github.com/esphome/feature-requests/issues/278

glmnet commented:
Looks great, unfortunately I also don't have the required hardware.

To help a little the testing souls:
once you copy the files into the custom_components folder this yaml should compile

````yaml
light:
  - platform: monochromatic
    name: "Kitchen Lights"
    output: output_component1

# Example output entry
output:
  - platform: dimmer
    id: output_component1
    gate_pin: GPIO5
    zero_cross_pin: GPIO20

````

HFsi commented:
I've tested the component with both RobotDyn and Ardutex module. Neither worked.
The RobotDyn has a PWM led onboard that blinks deppending the input, and it remains off regardless the state (full on or dimmerized).
Please let me know if I can help with some testing.

This is my config:
esphome:
  name: dimmer
  platform: ESP8266
  board: nodemcuv2

wifi:
  ssid: "xxx"
  password: "xxxxxxxxx"

  ap:
    ssid: "Dimmer Fallback Hotspot"
    password: "eUUqcUvJ0k1m"

captive_portal:

logger:

api:

ota:

light:
  - platform: monochromatic
    name: "Kitchen Lights"
    output: output_component1

output:
  - platform: dimmer
    id: output_component1
    gate_pin: D5
    zero_cross_pin: D2

glmnet committed:
fixes

basically missed the output pin setup and in the switching was switching true true true :P

glmnet committed:
fixes

basically missed the output pin setup and in the switching was switching true true true :P

glmnet commented:
@HFsi  would you like to try again? I just apply some patches, you need to update the cpp file only.

glmnet commented:
Also support for many outputs with same zc detector is expected

OttoWinter committed:
Format

OttoWinter committed:
Format

OttoWinter committed:
Enable ESP32

OttoWinter committed:
Enable ESP32

OttoWinter committed:
Also setup ZC pin

OttoWinter committed:
Also setup ZC pin

OttoWinter committed:
Support multiple dimmers sharing ZC pin

OttoWinter committed:
Support multiple dimmers sharing ZC pin

OttoWinter commented:
Ugh, turns out ESP32 implementation will need some rethinking.

The `timerStop` etc functions are all placed in flash, and therefore can't be called from the ISR.

Either we switch to esp-idf's timer APIs, or we use a timer that fires at a constant interval.

OttoWinter committed:
Fix ESP32

OttoWinter committed:
Fix ESP32

OttoWinter committed:
Lint

OttoWinter committed:
Lint

OttoWinter commented:
Ok, made some changes and works for ESP8266 and ESP32 now (used the following config to test; press the GPIO0 button on the board at a constant interval, and observe the LED):

```yaml
output:
  - platform: dimmer
    id: dimmer1
    gate_pin: LED
    zero_cross_pin:
      number: BUTTON
      mode: INPUT_PULLUP
      inverted: yes

light:
  - platform: monochromatic
    output: dimmer1
    name: Dimmer Light
```

HFsi commented:
Now works, but in dimmerized state, the intensity of the light isn't constant, and from time to time, flashes

glmnet commented:
![image](https://user-images.githubusercontent.com/96162/69471671-a1d09d80-0d80-11ea-96dc-4dc271978972.png)
D0 is an arduino trying to simulate a ZC (i have the maths wrong) and D2 is the pulse generated by the esp8266

@HFsi  can you give more details, can post full yaml? are you using the esp for something else?

minsuke commented:
you great job ! 
 I also test it, if I get Ac dimmer from amazon.

HFsi commented:
I posted a video on my YouTube, I think it'll be easier to show what is happening.
https://www.youtube.com/watch?v=tPHYKOM0F6c
My config is the same I've posted above, the dimmer is the only thing that controls the NMCU

glmnet commented:
Is there any % value for brightness where it will be dimmed and will be steady?

OttoWinter commented:
@HFsi Thanks for the video! Interesting that it appears to be quite periodic. Probably one of two things:

 - The disable signal is too late, and the TRIAC is not disabled before the next ZC pulse. The video shows ~gradular increase/decrease in brightness (as opposed to instant), but that might be due to the lamp itself.
 - The timing is off in another way, maybe the waveform interrupt is not as stable as we'd like.

HFsi commented:
At 144 of 255 the lamp turns off completely. In higher values seems more stable, but the perdiodic flash remains.
I've been struggling for a while with AC dimming, I've tested at least 5 different kinds/brands of lamps, so I can confirm that this one is the friendliest to dimm.
For example, this same lamp and module works like charm with this firmware:
https://github.com/nassir-malik/IOT-AC-Light-Dimmer-With-Alexa/blob/master/Part%20%23%202/Alex_mqtt_robodyn_dimmer/robodyn_dimmer.h
But is away of my understanding to see the difference between that one and the esphome one, I'm sorry I can't be on much help on that.
Of course, I'm willing to do all the testing that you guys may need

OttoWinter commented:
@HFsi Some thoughts:

 - "At 144 of 255 the lamp turns off completely" - part of that is certainly due to gamma correction. You can disable it by setting `gamma_correct: 0` in the light config.
 - "works like charm with this firmware:" - yeah I've seen a couple of implementations similar to that one, but the problem is that method will only be able to support 1 dimmer per ESP.

One difference I see from the one you posed is the ZC pin setup. If you set up the pin this way we can rule that out as a cause:

```yaml
output:
  - platform: dimmer
    id: dimmer1
    gate_pin: LED
    zero_cross_pin:
      number: GPIOx  # pin ZC is connected to
      # Inverted mode and pullup enabled
      mode: INPUT_PULLUP
      inverted: yes

light:
  - platform: monochromatic
    output: dimmer1
    name: Dimmer Light
```

If that's not it, I don't think you'll be able to help much anyway. This will need to be looked at with a proper function generator+oscilloscope.

HFsi commented:
These are the results with the changes you suggested.
Triac Dimmer:
https://www.youtube.com/watch?v=SR2sRZ9VEes
Mosfet Dimmer:
https://www.youtube.com/watch?v=vrst5RVQ4is

The best results I've reached (with any firmware) are using the Mosfet dimmer from ardutex.
I don't see much of a problem if the limitation is 1 dimmer per ESP, anyway a Sonoff Basic is about 5 bucks per dimmer, but the result would be an awesome dimmer, that is not so easy to get.

In any case, super great job! It's a BIG step

HFsi commented:
And the last test, a compare with a commercial dimmer vs ESP32 and Mosfet module
https://www.youtube.com/watch?v=7uw-cMg_-d0

glmnet commented:
Have been playing with the logic analyzer...
Modified function generator, a pro mini running this

````c++
void loop()
{
  // 50 hz sim
  delay(9);
  digitalWrite(13, HIGH);
  delay(1);
  digitalWrite(13, LOW);
}
````

the loaded robotdyn code linked by @HFsi 

Got this 
Robotdyn for brightness 1
![image](https://user-images.githubusercontent.com/96162/69507462-60bbc300-0f11-11ea-8762-c97ff0909ba5.png)
Robotdyn for brightness 100
![image](https://user-images.githubusercontent.com/96162/69507480-7630ed00-0f11-11ea-915b-630206c01025.png)

Robotdyn for brightness 220 (at 250 it seems to broke)
![image](https://user-images.githubusercontent.com/96162/69507487-79c47400-0f11-11ea-94f8-0ea64dedeee1.png)

ESP Home does this (ESP8266)
brightness little (slider close to the left)
![image](https://user-images.githubusercontent.com/96162/69507564-bdb77900-0f11-11ea-935a-8f84ce1da688.png)

brightness half
![image](https://user-images.githubusercontent.com/96162/69507616-f22b3500-0f11-11ea-90aa-c14c9e5b2f16.png)

brightness almost full
![image](https://user-images.githubusercontent.com/96162/69507693-269ef100-0f12-11ea-806c-07c4ce2998a6.png)

brightness about 10%
![image](https://user-images.githubusercontent.com/96162/69507763-5ea63400-0f12-11ea-9d27-90d88315bb24.png)

I'm not triac expert, nor do I know how the ZC detector really works.
I believe the code can be modified to mimic what robotdyn lib does, but I'm leaving on vacations for a week and wont be able to try it

HFsi commented:
Another info I can provide is the method of dimming.
There are Leading Edge and Trailing Edge. IMHK Triacs dimm by leading, and mosfets dimms by trailing, but I don´t really know if the ways are exclusive for the component. Maybe someone wiser can use that info.
The most perfect comercial dimmer I´ve ever seen, is the one from my last dimmer, and it has 2 mosfets

CSnipper commented:
With this code dimming not works - i I try dim, bulb turns off :(

```
output:
  - platform: dimmer
    id: dimmer1
    gate_pin: LED
    zero_cross_pin:
      number: GPIO0
      mode: INPUT_PULLUP
      inverted: yes

light:
  - platform: monochromatic
    output: dimmer1
    name: Dimmer Light
```

PWM to D4 (GPIO2)
ZC to D2 (GPIO4)

CSnipper commented:
Here are official RobotDyn libraries for different boards: https://github.com/RobotDynOfficial/RBDDimmer

Expaso commented:
I have posted a working custom integration for the ESP32 in the feature-request, without the use of interrupts.

Maybe I can help  to sort this thing out also for the ESP8266? (but we have less periphirals here)
I have a few robotdyn modules to test with (single and dual channel), and created firmware for triac-based timmers before.

glmnet commented:
@Expaso I've seen a zip file, could you create a PR?

Expaso commented:
Its a custom component right now, so Its meant to be in a custom_components folder in your esphome directory, and included there.

 I am not sure how I could PR this into the standard..

GregJArnold commented:
I've been playing with this for a while.  

My setup is a WeMos D1 mini (ESP8266), with a H11AA1 as the ZCD, and MOC3011s driving TRAICs for the output.  My ZCD goes high on zero crossing.

A couple of notes:

- Maybe have the ISR triggered on RISING if the pin's inverted, and FALLING if not?
- I also get the flickering at <100% brightness
- If I let the timer callback run freely, it calls itself too fast and the watchdog triggers a reset when its booting.  If I cap the timer callback to run no less than every 1000 microseconds, it'll boot.

I've done dimming on other stuff, and my way of handling it was to fire the gate however far into the cycle, leave it on the rest of the cycle, and then shut it off on the zero crossing.  This way, I just had a timer ISR running every 1/(60 * 2 * 256) seconds and it'd trigger whichever gates needed firing based on the dim level.  I was using an AVR, so I reset the cycle counter for the timer interrupt on zero crossing.

richrd commented:
Is anyone aware of some off the shelf flashable ESP dimmer modules that this could be used with?

CSnipper commented:
>     * "works like charm with this firmware:" - yeah I've seen a couple of implementations similar to that one, but the problem is that method will only be able to support 1 dimmer per ESP.


Works like a charm and controls two AC motors for my venitation system ;)

Expaso commented:
> Is anyone aware of some off the shelf flashable ESP dimmer modules that this could be used with?
Hmm, AC Dimmer modules such as Robotdyn should not be too hard to connect. It's only 3 wires. 
For real off the shelf solutions you'd better grab a zigbee or zwave dimmer device I suppose.

glmnet committed:
off gate on zc detect

glmnet committed:
off gate on zc detect

glmnet commented:
@HFsi  I changed so the gate is set low just on zc detect, this mimics what the robot dyn code you linked to does. Would you mind running tests again and give feedback?
Thanks.

glmnet committed:
tests pins validation

glmnet committed:
tests pins validation

glmnet committed:
Climate Mitsubishi (#725)

* add climate

* Mitsubishi updates

* refactor mitsubishi to use climate_ir

* lint

wilmardo committed:
fix: only decode when not str already (#923)

Signed-off-by: wilmardo <info@wilmardenouden.nl>

glmnet committed:
fix climate-ir bad merge (#935)

* fix climate-ir bad merge

* add mitshubishi test

Anonym-tsk committed:
http_request: fix memory allocation (#916)

Anonym-tsk committed:
http_request version fix (#917)

OttoWinter committed:
PID Climate (#885)

* PID Climate

* Add sensor for debugging PID output value

* Add dump_config, use percent

* Add more observable values

* Update

* Set target temperature

* Add autotuner

* Add algorithm explanation

* Add autotuner action, update controller

* Add simulator

* Format

* Change defaults

* Updates

timsavage committed:
Use b''.decode() instead of str(b'') (#941)

Handling of request arguments in WizardRequestHandler is not decoding
bytes and rather just doing a str conversion resulting in a value of
"b''" being supplied to the wizard code.

Valcob committed:
Adding the espressif 2.6.3 (#944)

Commit:
extract and use current version of python 3 (#938)

voibit committed:
Inverted output in neopixelbus (#895)

* Added inverted output

* Added support for inverted output in neopixelbus

* Update esphome/components/neopixelbus/light.py

Co-Authored-By: Otto Winter <otto@otto-winter.com>

* Update light.py

* corrected lint errors

Co-authored-by: Otto Winter <otto@otto-winter.com>

cyberplant committed:
Added degree symbol for MAX7219 7-segment display. (#764)

The ascii char to use it is "~" (0x7E).

Disclaimer: I didn't test this yet.

andrasbiro committed:
Fix dump/tx of 64 bit codes (#940)

* Fix dump/tx of 64 bit codes

* fixed source format

dmkif committed:
Update hdc1080.cpp (#887)

* Update hdc1080.cpp

increase waittime, to fix reading errors

* Fix: Update HDC1080.cpp

i fixed the my change on write_bytes

glmnet committed:
add tcl112 support for dry, fan and swing (#939)

panuruj committed:
Fix SGP30 incorrect baseline reading/writing (#936)

* Split the SGP30 baseline into 2 values

- According to the SGP30 datasheet, each eCO2 and TVOC baseline is a 2-byte value (MSB first)
- The current implementation ignores the MSB of each of the value
- Update the schema to allow 2 different baseline values (optional, but both need to be specified for the baseline to apply)

* Make both eCO2 and TVOC required if the optional baseline is defined

* Make dump_config() looks better

ayufan committed:
Add register_*_effect to allow registering custom effects (#947)

This allows to register custom effect from user components,
allowing for bigger composability of source.

timsavage committed:
Bugfix/normalize core comparisons (and Python 3 update fixes) (#952)

* Correct implementation of comparisons to be Pythonic

If a comparison cannot be made return NotImplemented, this allows the
Python interpreter to try other comparisons (eg __ieq__) and either
return False (in the case of __eq__) or raise a TypeError
exception (eg in the case of __lt__).

* Python 3 updates

* Add a more helpful message in exception if platform is not defined

* Added a basic pre-commit check

kbx81 committed:
Add transmit pioneer (#922)

* Added pioneer_protocol to support transmit_pioneer

glmnet committed:
Display tm1637 (#946)

* add TM1637 support

mario-tux committed:
Support a further variant of Xiaomi CGG1 (#930)

Fall3n-be commented:
With this firmware instead of dimming it just starts flickering more, I'm working with a ESP8266.
`light:
  - platform: monochromatic
    name: "Dimmer test"
    output: output_component1

output:
  - platform: dimmer
    id: output_component1
    gate_pin: D5
    zero_cross_pin: D2`

Fall3n-be commented:
With this firmware on an ESP8266 instead of dimming it just starts flickering more.
```light:
  - platform: monochromatic
    name: "Dimmer test"
    output: output_component1

output:
  - platform: dimmer
    id: output_component1
    gate_pin: D5
    zero_cross_pin: D2```

hectorgimenez committed:
Daikin climate ir component (#964)

* Daikin ARC43XXX IR remote controller support

* Format and lint fixes

* Check temperature values against allowed min/max

LegendOfSparta commented:
Hi again, I’m sorry for being away lately, anyways after many many circuit connections the results were satisfying, so i tried to just use this Ardunio IDE code https://github.com/nassir-malik/IOT-AC-Light-Dimmer-With-Alexa/tree/master/Part%20%23%201/robodyn_dimmer
The results were too perfect! 

I wanna use this “hw-timer” library with esphome, is it possible!? like integrating the hw timer library in my esphome configuration and use lambda or something to get this working?!

Svrbinek commented:
Hi, can you please tell me how you started Nasir Malik firmware? I tried it too, but at least it is possible to choose 1% and that ordinary light bulb still faintly filament. Then I didn't figure out how to turn the bulb off completely, and then if I chose control number 50 in the TV series, it flashes briefly for a moment and then lights up full and doesn't dim until I remove the bulb from the socket and return it again. But it is so. I have a ESP8266 board and downgrad to 2.3.0

glmnet committed:
fix tm1637 missing __init__.py (#975)

glmnet committed:
Merge branch 'ac-dimmer' of https://github.com/esphome/esphome into ac-dimmer

zaibas01 commented:
hi 
I made two dimers: trailing and  leading edge. Both work well. But lamp turn on when brightness 50 percent. On both dimmers the same problem.
[dimeriu schemos.pdf](https://github.com/esphome/esphome/files/4198294/dimeriu.schemos.pdf)
[https://www.youtube.com/watch?v=ViW1zfnVw9A](url)

glmnet commented:
> hi
> I made two dimers: trailing and leading edge. Both work well. But lamp turn on when brightness 50 percent. On both dimmers the same problem.
> [dimeriu schemos.pdf](https://github.com/esphome/esphome/files/4198294/dimeriu.schemos.pdf)
> https://www.youtube.com/watch?v=ViW1zfnVw9A&t=15s

Did you test it on ESP8266 or ESP32 only?

I've noticed similar behavior with the 8266, it changes somewhat if you adjust the `gamma_correction` but it does not fixes it properly, may be a `minimum_on` setting is required like some wall dimmers which uses as trim-pot

I'm only testing leading edge so far (TRIAC)

zaibas01 commented:
i test it only on ESP32,.  
you think there is a difference?. i can try with ESP8266 too

glmnet commented:
The implementations are quite different, there are users reporting flickering but then I pushed a change for the ESP8266 and it's somewhat working for me, no flickering but dim levels are no lineal with value yet. Only tested TRIAC and only one lamp.
MOSFET method is not implemented for the ESP8266, I still have to play around with ESP32.

zaibas01 commented:
tested on esp8266. work well,  no flickering but the same problem with  brightness

glmnet commented:
Ahh thanks for the feedback, I'll work on the `minimum_on` setting then. I'm fully open to suggestions on configuration method, parameter naming etc.

Aside of the dim range, I've found out my lamp will turn on just at 50% but then I can decrease the brightness and it will be still be lit, i.e.
0% -> OFF
30% -> OFF
50% -> ON, dimmed
30% -> ON, less dimmed

Has anyone seen this behavior too? I was wondering what will happen if I send to the lamp one or two full cycles of AC when turning on and then start the phase cut

Svrbinek commented:
Hi, I'd like to test too. I have ESP32 and ESP8266, but only a standard 60W tungsten lamp. At that I found that using FW from nassir-malik there are problems because she fails to work from more than 10% dimming. Above this level, the output stage opens completely and it is possible to observe the course on an oscilloscope. Can you test your solution on an ordinary light bulb?

zaibas01 commented:
> Ahh thanks for the feedback, I'll work on the minimum_on setting then. I'm fully open to suggestions on configuration method, parameter naming etc.

Aside of the dim range, I've found out my lamp will turn on just at 50% but then I can decrease the brightness and it will be still be lit, i.e.
0% -> OFF
30% -> OFF
50% -> ON, dimmed
30% -> ON, less dimmed

Has anyone seen this behavior too? I was wondering what will happen if I send to the lamp one or two full cycles of AC when turning on and then start the phase cut

0% -> OFF
30% -> OFF
52% -> ON, dimmed
47% -> off, less dimmed

zaibas01 commented:
my test:
0% -> OFF
30% -> OFF
53-55% -> ON, dimmed
45-47% -> off, less dimmed

GregJArnold commented:
The reason it exhibits the dimming behavior where it won1t glow at about 30% ramped up from zero but will when ramped down is to to the physical properties of the filament - it has a *negative*  temperature coefficient of resistance - as it gets hotter, its resistance goes *up*.  So, what's happening is the low current can't heat up a filament, but it can keep a hot one up to temperature.  This is a known behavior of dimming incandescent bulbs, and manually-controlled dimmers show this effect.

glmnet commented:
Alright that makes sense however I’m experiencing with led  driven lamps only

glmnet committed:
fix esp8266 many dimmers, changed timing

glmnet commented:
I've made a change in the timing 
Testing with ESP8266 only, but should affect ESP32 too.
Fixed frequency of mains being reported twice as fast (100hz instead of 50hz here)
Fixed issue with attach_interrupt crashing esp8266
Changed timing to always enable at least 1/4 of the wave form, now my lights both turn on at 28, I have a dimmable LED and a 12V halogen lamp with an electronic transformer, both can dim simultaneously each at its own level sharing the ZC pin.

No flickering but occasionally seems like a lamp is turning on when it shouldn't, didn't find a pattern of this

zaibas01 commented:
bulb turns on 11%, but  now can't set lowest brightness

puneit-hassio commented:
Can this also be used to control the speed of a ceiling fan?

glmnet commented:
> bulb turns on 11%, but now can't set lowest brightness

so are you telling me that a lower dim level was possible before?

zaibas01 commented:
in my case, yes

Luisiado commented:
with this conf and d1 MINI
```
output:
  - platform: dimmer
    id: dimmer1
    gate_pin: D7
    zero_cross_pin:
      number: D6
      mode: INPUT_PULLUP
      inverted: yes

light:
  - platform: monochromatic
    output: dimmer1
    name: Dimmer Light
```

I can switch ON and OFF the bulb, I have also flickering(100%). The funny point is that I can not regulate the light intensity the component works like ON/OFF switch. I tried the robotdym library and works ok with the same setup.

I also tried ESP32 and the same problem.

glmnet commented:
@Luisiado can you check the logs? What line hertz is in your zone and what the logs says? It might not be detecting ZC properly

Luisiado commented:
@glmnet my frequency is 50Hz
I am using RobotDyn-RBDDimmer

´´´
[12:49:04][V][dimmer:151]:   Estimated Frequency: 49.915Hz
[12:50:30][VV][api.service:311]: on_light_command_request: LightCommandRequest {
  key: 1022277722
  has_state: YES
  state: YES
  has_brightness: YES
  brightness: 0.521569
  has_rgb: NO
  red: 0
  green: 0
  blue: 0
  has_white: NO
  white: 0
  has_color_temperature: NO
  color_temperature: 0
  has_transition_length: NO
  transition_length: 0
  has_flash_length: NO
  flash_length: 0
  has_effect: NO
  effect: ''
}
[12:50:30][D][light:264]: 'Dimmer Light' Setting:
[12:50:30][D][light:277]:   Brightness: 52%
[12:50:30][D][light:303]:   Transition Length: 1.0s
[12:50:30][VV][api.service:104]: send_light_state_response: LightStateResponse {
  key: 1022277722
  state: YES
  brightness: 0.521569
  red: 0
  green: 0
  blue: 0
  white: 0
  color_temperature: 0
  effect: ''
}
[12:50:30][VV][preferences:051]: SAVE 34: 0=0x00000001 1=0x3F058586 (Type=1022277722, CRC=0x0744ED6D)
´´´
and according to the first table of the document I can use D6 as zero_cross_pin and D7 as output or gate.

Svrbinek commented:
I just tested it with an oscilloscope. It seems that if it is 0-10% it is off and no signal is generated even on the oscilloscope. At 11% it switches up to 99% without changing the waveform. It is displayed with only one short trip pulse per 100Hz wave. 100% turns on completely. No change is noticeable on the bulb. Only on / off.
Light off 0%
![20200226_170355](https://user-images.githubusercontent.com/60275636/75364182-9038ed80-58bb-11ea-8efd-3d344928f80b.jpg)
Light dimming 10%
![20200226_170049](https://user-images.githubusercontent.com/60275636/75364188-9202b100-58bb-11ea-908b-d51c0a9734be.jpg)
Light dimming 11% (11-96%)
![20200226_170126](https://user-images.githubusercontent.com/60275636/75364191-929b4780-58bb-11ea-8904-6b8a6c0f25c3.jpg)
Light dimming 11% - detail
![20200226_170153](https://user-images.githubusercontent.com/60275636/75364194-93cc7480-58bb-11ea-9b30-c4ad2b43a535.jpg)
Light dimming 97% (1. change 97-99%)
![20200226_170210](https://user-images.githubusercontent.com/60275636/75364195-94fda180-58bb-11ea-84fd-40b5219d7c17.jpg)
Light on 100%
![20200226_170322](https://user-images.githubusercontent.com/60275636/75364196-95963800-58bb-11ea-9cd0-09b7a61ab8f5.jpg)

glmnet commented:
@Svrbinek this is useful, can you review the captions on the pictures are accurate, e.g. 2nd and 3rd looks the same for me, otherwise can you explain the difference? Thanks.

glmnet commented:
As an informational note, I've played with the RoboDyn again this time with the real hardware, and I found out it works better, specially dimming at low brightness.
I don't know why @Svrbinek cannot get proper dimming in between 11%-99% it is working for me and also for @zaibas01

Svrbinek commented:
Captions are for each photo, how many percent it was taken and what it shows. If you write what you do not understand exactly and I will try to clarify. I just tried this one (nassir-malik
/
IOT-AC-Light-Dimmer-With-Alexa) and only with a classic filament lamp and behaves strangely. 60W bulb dims from 0-10% and at 11% or more jumps to 100% after a while, even if the PWM signal does not give an impulse (probably thyristor can not handle anything with the power at the end). 25W shines from 1-100% full. 75W from 1-35% dims and then jumps to the full. But if I connect 75W and 25W in parallel, so it works from 1-99% dimming as it should and 100% then goes full, ie without error. I have no idea why performance affects the course and control of the thyristor.

Svrbinek commented:
Now I noticed that there is a mistake in the description of the photos. Photos uploaded shifted, so the first photo is actually 100%. I'll try to fix the labels

Svrbinek commented:
I also attach photo of my ESP8266 and the code I use in it.
![20200226_201822](https://user-images.githubusercontent.com/60275636/75379286-6e4c6480-58d5-11ea-9516-7fbab5e8971e.jpg)

output:
  - platform: dimmer
    id: dimmer1
    gate_pin: GPIO5
    zero_cross_pin:
      number: GPIO4
      mode: INPUT_PULLUP
      inverted: yes

light:
  - platform: monochromatic
    output: dimmer1
    name: Dimmer Light

glmnet commented:
Can you post image of the dimmer? Triac and ac stuff

Svrbinek commented:
Sure, here are the bulbs and the triac module
![20200227_175839](https://user-images.githubusercontent.com/60275636/75466786-00627480-598b-11ea-8c6e-b5c15100030b.jpg)
![20200227_175844](https://user-images.githubusercontent.com/60275636/75466793-022c3800-598b-11ea-82fe-84e7c5443300.jpg)

glmnet committed:
Increased value resolution, added min power

glmnet committed:
use min_power from base class

glmnet committed:
fix min_power. add init with half cycle

glmnet commented:
Thanks all for your support!

I've just added three changes, 
1. Increased resolution of internal brightness, since the light brightness goes thru a gamma correction thing, so
   brightness, float:  0-100%
   gamma, float: 0-100% but values like 10% are now 0.05, etc.
   to internal value 0-255, so now many low values (from 0 to 30%) ends in the 0-1 range, value lost
   now internal value goes to 0-65535 precision, and seems to work better.

2. Kind of obey float output's `min_power`, which defaults to 10% for the dimmer now, there is a minimum absolute required cycle time to not have a range of no turn on at all.

3. Added `init_with_half_cycle` which will cause the first half ac cycle to be sent complete. This helps the dimmable electronic led lights controller to power on, they will not yet turn on the LEDs but will be able to start from cold to a low brightness level, this one defaults to true as most lamps are LED now, but you might want to set it to false on tungsten lamps or if you experience a flickering at the turn on.

so you can play with the settings and please give feedback.

My configuration: **see min_power**
````
output:
  - platform: dimmer
    id: dimmer1
    gate_pin: D7
    min_power: 15%
    init_with_half_cycle: false
    zero_cross_pin:
      number: D6
      mode: INPUT
      inverted: yes
````

Also you might want to play with `gamma_correction` too, check the [light docs](https://esphome.io/components/light/index.html)

I don't have a scope but looks great at the logic analyzer:
0%
![image](https://user-images.githubusercontent.com/96162/75497179-3ced9a80-59a2-11ea-9bed-46300336c0d7.png)
5%
![image](https://user-images.githubusercontent.com/96162/75497222-5bec2c80-59a2-11ea-996f-0415d3ccfabb.png)
95%
![image](https://user-images.githubusercontent.com/96162/75497268-85a55380-59a2-11ea-8cdc-1fb7decf1f73.png)
100%
![image](https://user-images.githubusercontent.com/96162/75497289-9786f680-59a2-11ea-808d-c20e0d3e4ce3.png)


Disclaimer: only tested on ESP8266, two dimmers simultaneously sharing the same ZC pin, the TRIACs are BT139 600E

Svrbinek commented:
Hello, I have a question about how to get the best edited files into Home Assistant. I have always copied the code, pasted it into a text editor and saved it. Then he edited the name and uploaded it to HA. But it is possible that there may be an error in this way, because the dimmer.cpp file in HA shows only one line with some characters. Download the whole esphome folder with all the codes to different devices I don't want. Thanks for the help.

Svrbinek commented:
So I tried to copy the new codes and paste them directly into HA using the File Editor, but it's unchanged. I do not know where the error is, but the light either shines or not. Now it lights up to 100% from 2% in HA :-(

Svrbinek commented:
Can I ask for help where I can be wrong? I have files in HA saved in this path "/config/esphome/custom_components/dimmer". I'll send the screen. I made the settings in esp8266 exactly according to the latest glmnet update.
![Image 1](https://user-images.githubusercontent.com/60275636/75572702-91088580-5a5b-11ea-92ce-2650368b8326.jpg)

glmnet commented:
Hmmm, I messed it up somehow, there are 55 files changes, there should be just 4.
Anyway you can tell you have the latest files if in your log you see

`  dimmer: Init with half cycle: YES`

also this code should be at the end of the cpp

Svrbinek commented:
Here I recorded a video of how it behaves on an oscilloscope. Light off and then on 10%
https://youtu.be/wPWu0WXE49k

zaibas01 commented:
> I just tested it with an oscilloscope. It seems that if it is 0-10% it is off and no signal is generated even on the oscilloscope. At 11% it switches up to 99% without changing the waveform. It is displayed with only one short trip pulse per 100Hz wave. 100% turns on completely. No change is noticeable on the bulb. Only on / off.

replace  zero crossing optocoupler with H11L1. 
[dimmer.pdf](https://github.com/esphome/esphome/files/4270850/dimmer.pdf)

nliaudat commented:
Act only as switch. 
The waveform on zero_cross pin is alway the same
No waveform nor signal on gate pin

#hardware : 
robodyn  ac dimmer+ lolin nodemcu v3

Regards

Svrbinek commented:
The mistake must be somewhere in the code, because it is exactly how it is written. It works with another code. I use this model https://robotdyn.com/ac-light-dimmer-module-1-channel-3-3v-5v-logic-ac-50-60hz-220v-110v.html

zaibas01 commented:
> Disclaimer: only tested on ESP8266, two dimmers simultaneously sharing the same ZC pin, the TRIACs are BT139 600E


what kind of optocoupler on zero_crossing input?

glmnet commented:
@zaibas01 

this is my board, I can read PC817 on the optocoupler connected to the ZC detector, the board is powered by the 3.3v of the esp development board.

The Mosfet drivers seems to say MOC3203, and as I say TRIACs are BT139 600E

![IMG_9353](https://user-images.githubusercontent.com/96162/75636236-b16f4600-5bfb-11ea-8e9a-18af7eb9e741.jpg)

zaibas01 commented:
> this is my board, I can read PC817 on the optocoupler connected to the ZC detector, the board is powered by the 3.3v of the esp development board.

The Mosfet drivers seems to say MOC3203, and as I say TRIACs are BT139 600E

it's a different scheme. For  zero crossingit uses a transistor BC847 to amplify the signal. Robotodyn module for zero crossing uses only 4n25. I made own dimmer according to robotdyn scheme. Dimer did not work. 
Then i replaced 4n25 with H11L1. Now everything works!!!  

Be careful 4n25 and H11L1 pinout is different.  You also need to connect + 3.3-5V to  H11L1 6 pin. Scheme attached
[dimmer.pdf](https://github.com/esphome/esphome/files/4275243/dimmer.pdf)


Sorry for my english :)  
if you have any questions ask.

zaibas01 commented:
* However, the ESP32 input-pins do not have schmitt-trigger buffers, which means they are very sentitive to hysteresis.
 * This can lead to flicker or other unwanted effects, when used with the robotdyn dimmer or other dimmers without schmitt-trigger outputs.
 * To overcome this, please buffer the zero-crossing detection pin using a schimitt-trigger buffer like the 74LVC3G17DP, 
 * or, replace the opto-coupler on the dimmer (which is probably someting like the 4N25) with a H11L1 opto-coupler (with build-in schmitt trigger).

glmnet commented:
@zaibas01 just to make it crystal clear, you got my code working on your setup?
Are saying that RobotDyn and my board are different but what really matters is the optocoupler on the Zero Cross side?

Anyway there must be something going on at the RobotDyn code so it works with RobotDyn boards on the ESP8266 with the 4n25, may be how the interrupt is setup

zaibas01 commented:
> @zaibas01 just to make it crystal clear, you got my code working on your setup?
Are saying that RobotDyn and my board are different but what really matters is the optocoupler on the Zero Cross side?

Anyway there must be something going on at the RobotDyn code so it works with RobotDyn boards on the ESP8266 with the 4n25, may be how the interrupt is setup


https://www.youtube.com/watch?v=ViW1zfnVw9A    in video Trailing-edge Dimmer, but i made and Leading-edge dimmer (TRIAC dimmers) according to robotdyn scheme. 
The result is the same. 
![dimmer](https://user-images.githubusercontent.com/46297764/75678196-c13b6880-5c95-11ea-8779-693d12e28a9e.jpg)    
your scheme zc is a little different (scheme attached). 
it uses the BC847 transistor, maybe a signal amplifier. I'm not an electronics  so I can't tell you exactly.
![zc](https://user-images.githubusercontent.com/46297764/75679761-e7163c80-5c98-11ea-9a3f-53ac53ad51d0.jpg)

glmnet commented:
Hmm, thanks for the video, it should turn on before, did you try different values of `min_power`?
does it behave the same with the triac dimmer?

I could add a config for trailing edge dimming if you want to test later.

zaibas01 commented:
> Hmm, thanks for the video, it should turn on before, did you try different values of min_power?
does it behave the same with the triac dimmer?

I could add a config for trailing edge dimming if you want to test later.

video with old dimmer code.  
I haven't tried with a new  yet

zaibas01 commented:
glmnet i can solder and send both dimmers to you. 
Where are you from?

Svrbinek commented:
glmnet I'm willing to sponsor the original robotdyn module to program it directly on it. I doubt that people will modify the module they buy. It does not make any sense. Now I understand why the RobotDyn module doesn't work for this code. It's quite something else. Will it be possible to program this to this module? https://robotdyn.com/ac-light-dimmer-module-1-channel-3-3v-5v-logic-ac-50-60hz-220v-110v.html

glmnet commented:
Of course I'll accept donations, contact me at glm.net @ gmail

glmnet commented:
So looking around my stuff it happens I have a 4n35 optocoupler and I should have a bridge rectivier too...
Do someone has the RobotDyn schematic?

It's not the same optocoupler but seems close enough

zaibas01 commented:
![drawing](https://user-images.githubusercontent.com/46297764/75705220-9f58da80-5cc3-11ea-8452-cdec70be95a5.jpg)
![Schematic](https://user-images.githubusercontent.com/46297764/75705222-a08a0780-5cc3-11ea-8779-b11b41776e4f.jpg)

glmnet commented:
what would be the value of R1 and R3?

glmnet committed:
added method for trailing pulse, trailing and leading

glmnet commented:
Hey folks,

I was thinking maybe the issue is not the opto but the triacs, mine might disconnect easily on low voltage but yours not, and so, since I keep the gate output on until the ZC that might prevent the triac going off.

so I implemented the `method` setting, valid values are `trailing pulse` (default) `trailing` (what we were using so far) and new bonus `leading` for those who wants to try dimming with mosfet, I don't have a mosfet dimmer so I don't know how it works, but judging the logic analyzer it should be ok.

here is the plot, ZC, trailing pulse, trailing and leading 33, 50 and 66% values resp.
![image](https://user-images.githubusercontent.com/96162/75734006-8d803300-5cd5-11ea-97f1-ebb70a8de8b3.png)

example config:
````yaml

output:
  - platform: dimmer
    id: dimmer1
    gate_pin: D7
    min_power: 12%
    init_with_half_cycle: false
    zero_cross_pin:
      number: D6
      mode: INPUT
      inverted: yes
  - platform: dimmer
    id: dimmer2
    gate_pin: D5
    method: trailing
    zero_cross_pin:
      number: D6
      mode: INPUT
      inverted: yes
  - platform: dimmer
    id: dimmer3
    gate_pin: D4
    method: leading
    zero_cross_pin:
      number: D6
      mode: INPUT
      inverted: yes

light:
  - platform: monochromatic
    output: dimmer1
    name: Dimmer Light Halogen
    id: l1
    gamma_correct: 0
  - platform: monochromatic
    output: dimmer2
    name: Dimmer Light LED
  - platform: monochromatic
    output: dimmer3
    name: Dimmer Light Leading
````

glmnet commented:
And I got the method names confused, of course

zaibas01 commented:
![zcd](https://user-images.githubusercontent.com/46297764/75748619-9c440580-5d28-11ea-91bb-430fef90674e.jpg)

zaibas01 commented:
> so I implemented the method setting, valid values are trailing pulse (default) trailing (what we were using so far) and new bonus leading for those who wants to try dimming with mosfet, 


Leading-edge dimmers (TRIAC dimmers) robotdynd dimmer
Trailing-edge Dimmers (Reverse phase dimmers) mosfets dimmer

nliaudat commented:
Thank you very much !
TRAILING_PULSE method working great with robotdyn ac dimmer

Svrbinek commented:
It still doesn't work for me. It behaves differently than before but still does not do what it has. I uploaded new files, edited yaml and restarted esphome. Now I will try to restart the whole HA.

Svrbinek commented:
So I tried the trailing method, partially dimming, but the bulbs blink and still flash. The setting is 50%. https://youtu.be/ssVEhgokfGM

nliaudat commented:
> 
> 
> It still doesn't work for me. It behaves differently than before but still does not do what it has. I uploaded new files, edited yaml and restarted esphome. Now I will try to restart the whole HA.

1) Do you have a the latest version of esphome on your workstation ? (I've seen your screenshots and I mean you are using hass.io version of esphome.)
2) Please test the 3 methods : TRAILING_PULSE, TRAILING and LEADING

Anyway I have differents results for robodyn : 
    TRAILING_PULSE = work for incandescant lamp not for fan
    TRAILING = blink incandescant lamp, do not work for fan
    LEADING = do not work with lamp nor fan

I mean it could help
Regards

Svrbinek commented:
I have now tried all three methods and two of them are close to working. Dimming LED and Hallogen. But they both blink. Each different. The oscilloscope shows that the signal has an unstable and strange frequency. Zero crossing is set to 100Hz, but the PWM control is "random". I do not know what the frequency is set, but it should come into each zinc crossing crossing to avoid blinking.

brano1990 commented:
I tried TRAILING_PULSE = dimming doesn't work only on / off
            TRAILING = dimming works from 50%, but bulb flashes (stroboscope)

Can't there be a problem with our 50Hz network? your code is on 60Hz.
I´m using dimmable LED bulb and halagen and ESPHome 1.14.3 and I use HassOS 3.11 just like @Svrbinek - pozdravujem tiez stym zapasim

Svrbinek commented:
If the code is written for 60Hz network then it will be the problem. Is it possible to modify the code even for the 50 Hz network? :-) Thanks

Svrbinek commented:
Here you can see from the logic analyzer that it really jumps randomly. it will probably be 60Hz in the program, but only 50Hz on the network.
![Image 2](https://user-images.githubusercontent.com/60275636/75916810-4753dc80-5e59-11ea-90b8-f7d221b21702.jpg)

glmnet commented:
Thanks for the feedback!
50 or 60 Hz should work, every cycle the cycle time is calculated, you can check on the logs it will tell your mains frequency 

I’ll definitely need the different hardware, the logic analyzer output is actually telling me a few things, most importantly that the signals are all seen by the esp (the rising edge is in every zc for D4)

There might be two reasons 
1. The length of the pulse is different than mine
2. The esp is not seeing a clean pulse, like two or more hits. 

@svrbinek can you post that data in the logic analyzer format? Zipped?

glmnet committed:
fix method name. try filter invalid falling pulse

glmnet commented:
well, fixed the name I got wrong leading <-> trailing

also I attempt to compensate for invalid pulses if the esp is interrupted more than once for the same zc signal, I say attempt because I did not try to simulate this (I could but it takes more time)
So please as usual give feedback.
Thanks

brano1990 commented:
it works great 

it start dimming from 25% to 90%, above 90% sometimes it flashes, it´s stable from 25% to 90%
where can I adjust the percentage range? That the dimming starts from 10% and ends with 90% = 100%

Svrbinek commented:
It still doesn't work for me. Here is the online output on the oscilloscope all 4 channels. It can be seen that all 3 outputs still have some random pulses or breaks. 
Brano1990 please upload video?
https://youtu.be/cAFAM1LGrDI

brano1990 commented:
watch the video is seen at 100% flash twice Do not forget to upload a new dimmer.h dimmer ....
and upload esphome
https://youtu.be/WVPqf_lrSpI

"@Svrbinek funguje to ale iba od 25% do 90% potom to blikne ked pozres video nezabudni daj updada aj v esphome"

Svrbinek commented:
brano1990 can we get together? Somehow I do not know what I have wrong and do not want to spam here on the thread. Please email me at jezink (at) seznam dot cz

Svrbinek commented:
glmnet until now I have noticed that you requested a zip .. here is the current zip file
[24 MHz, 48 M Samples [5].zip](https://github.com/esphome/esphome/files/4300154/24.MHz.48.M.Samples.5.zip)

glmnet commented:
@Svrbinek 

![image](https://user-images.githubusercontent.com/96162/76132553-4a360500-5ff2-11ea-9900-7b31a6b9796e.png)
When I zoom in there I see this...
![image](https://user-images.githubusercontent.com/96162/76132575-5fab2f00-5ff2-11ea-987c-46d9bacd03c1.png)

Now the code I uploaded before should account this issue but seems not to be working ok. I'd have to reproduce this and test again, maybe I can quickly program an arduino uno to mimic this behavior.

You can try to put a small cap to see if it helps, like a 100nF between the ZC signal and the GND I really don't know much of electronics so if you have you can try other values too.

Svrbinek commented:
I think this is the wrong logic analyzer, it's not on the oscilloscope. Today I'm gonna record it from an oscilloscope. Then I try to record a waveform from Nasir Malik from a logic analyzer. Can't you find in the code with him how he has it?
https://github.com/nassir-malik/IOT-AC-Light-Dimmer-With-Alexa/tree/master/Part%20%23%202/Alex_mqtt_robodyn_dimmer

Svrbinek commented:
Here is the same on the oscilloscope that was on the analyzer. Here the errors are not seen in the ZC. Those errors are caused by a cheap analyzer.

https://youtu.be/M8YSG82kwX0

zaibas01 commented:
I did three tests.:
1. ESP32,   zc_detector optocoupler 4n25, triac BT139
2. ESP32,   zc_detector optocoupler 4n25, triac BTA16
3. ESP32,   zc_detector optocoupler H11L1, triac BT139 or BTA16 

https://www.dropbox.com/sh/ynaz47ewyufj67v/AAAK8D0NHhSEGexhczdk820Za?dl=0

output:
  - platform: dimmer
    id: dimmer1
    gate_pin: GPIO23
    min_power: 30%
    method: leading   # (did not work without it)
    init_with_half_cycle: false
    zero_cross_pin:
      number: GPIO22
      mode: INPUT
      inverted: yes

light:
  - platform: monochromatic
    output: dimmer1
    name: Dimmer Light bandymas 5    
    id: dimmer_light_bandymas_5

zaibas01 commented:
did a couple tests:
1 ESP32, zero crossing optocoupler 4N25,  triac  BTA16 or BT139 (there is no difference)

output:
  - platform: dimmer
    id: dimmer1
    gate_pin: GPIO23
    min_power: 25%
    method: leading   # (does not work without this)
    init_with_half_cycle: false
    zero_cross_pin:
      number: GPIO22
      mode: INPUT
      inverted: yes

light:
  - platform: monochromatic
    output: dimmer1
    name: Dimmer Light bandymas 5    
    id: dimmer_light_bandymas_5

result: https://www.youtube.com/watch?v=KUiP3FdtZLI

2 ESP8266, zero crossing optocoupler 4N25,  triac  BTA16 or BT139 (there is no difference)

output:
  - platform: dimmer
    id: dimmer1
    gate_pin: D6
    min_power: 20%
    method: leading  # (does not work without this)
    zero_cross_pin:
      number: D5
      mode: INPUT
      inverted: yes

light:
  - platform: monochromatic
    output: dimmer1
    name: Dimmer Light bandymas 5    
    id: dimmer_light_bandymas_5

result:  https://www.youtube.com/watch?v=qrWMD1jvmHY

Svrbinek commented:
Hello. So I have to announce that it finally works somehow. The error was possibly in poorly saved files, which helped brano1990. I have tested all 3 options from glmnet and the "halogen" setting with an ordinary light bulb works almost perfectly. It dims from 0 to 100% just it's not a linear waveform but it doesn't matter. Almost 70% shines full.
I have also tested the last 3 settings from zaibas01 and none of them go right there. If the bulb exceeds 60%, it will flash and light up fully, and then it will no longer be possible to turn it off.
Here is a waveform on an oscilloscope with a bulb
https://youtu.be/VVRx4Cyq6wQ
Here is the waveform itself
https://youtu.be/SL1l_bjSRzA

So I want to thank everyone for cooperation and glmnet programmer :-)

glmnet commented:
Thanks all for feedback. I’m glad it starting working for everyone 
Try gamma setting  
That should change how the Brightness correlates to  pwm length 
If that is not enough I can add another setting. May be max power, I’m open to suggestions of course

Svrbinek commented:
I have set gamma_correct: 0 and works well from 0 to 70%, then it shines almost full and the change is not noticeable on the bulb, only the oscilloscope shows the shift. I tried to change the gamma to a different value and that was all wrong. I don't know how to change this so that dimming expands to visible change up to 100%

GregJArnold commented:
I tried the most recent code, and it crashes whenever I try to turn the light on or off - if I have it dimmed to anything in between 0 and 100%, it gets caught in a boot loop, and I have to pull the Wemos from the socket to get it to quit boot looping.

`
light:
  - platform: monochromatic
    name: "Test Light"
    output: output_component1

output:
  - platform: dimmer
    id: output_component1
    gate_pin: D5
    min_power: 10%
    method: leading
    init_with_half_cycle: false
    zero_cross_pin:
      number: D2
      mode: INPUT
      inverted: no
`

glmnet commented:
can you provide usb logs?

GregJArnold commented:
Here you go.
[log.txt](https://github.com/esphome/esphome/files/4304518/log.txt)

glmnet commented:
I see you have a 60hz mains and your log says "wdt reset" so should be it... anybody else running successfully at 60hz?

AndresL71 commented:
@zaibas01
Did you replace the 4n25 with a h11l1 in a robotdyn dimmer ?? If so can you point out how I should modify the pins to make it work?? Can’t seem to. ...

zaibas01 commented:
> @zaibas01
Did you replace the 4n25 with a h11l1 in a robotdyn dimmer ?? If so can you point out how I should modify the pins to make it work?? Can’t seem to. ...

cant you try  witch esp32? for me it worked fine without modification, with 4N25

Svrbinek commented:
Hello friends. Today I got a dimmable source from China, which I decided to connect to my T8 LED tubes and test. Where the classic Dimmer Light Hallogen bulb worked, so this source does not work, but I tried other options from 3. Dimmer Light LED is also inoperative, because there is a blink. The only functional option is now Dimmer light leading, for which I am glad. True, it works up to about 45% -48% but it doesn't matter, because then it works up to 100%. Perhaps it will help someone. It is these resources https://de.aliexpress.com/item/32928524331.html?spm=a2g0s.9042311.0.0.567f4c4dKd9Bzz

CSnipper commented:
I tried this code with my 230V/50Hz AC motor for ventilation, it turns on at 60% and working max on 100%. 

Is it possible to change the work from 0 to 100%?

My code:

```
output:
  - platform: dimmer
    id: dimmer1
    gate_pin: D2
    min_power: 10%
    method: leading
    init_with_half_cycle: false
    zero_cross_pin:
      number: D6
      mode: INPUT
      inverted: yes

light:
  - platform: monochromatic
    output: dimmer1
    name: Wywiew
    id: wywiew
```

glmnet commented:
did you try increasing `min_power`

CSnipper commented:
OK, works perfect when min_power: 50% Thanks!

glmnet commented:
Can everybody report back if this thing is working/not working for you?

@GregJArnold  are you still having issues? I can try to figure out how to relax the WDT in your case

Also I haven't tested in ESP32 

Next step will be writing docs so it can be merged to dev 🎉

glmnet committed:
renamed to ac_dimmer

glmnet commented:
In case anyone needs info docs are temporarily available here https://deploy-preview-536--esphome.netlify.com/components/output/ac_dimmer.html

GregJArnold commented:
Yeah, I'm still having the same issue.  I assume that with a 60 cps line frequency, the watchdog isn't getting fed enough.

glmnet commented:
@GregJArnold can you post your full yaml file? I see you have dallas configured too

glmnet commented:
![pic_10_1](https://user-images.githubusercontent.com/96162/78948604-543e9e00-7a9f-11ea-99d2-196921375347.png)

Running at 60hz (simulated) no issues here

glmnet committed:
fix ESP32 not configuring zero cross twice

Svrbinek commented:
glmnet Hello. Have you got an AC dimmer robotdim by post?

glmnet commented:
> glmnet Hello. Have you got an AC dimmer robotdim by post?

not yet, Argentina borders are closed due to the thing and I need to get out to grab it

djares01 commented:
I’ve problems with the dimmer, I have a robotdyn with ESP32 and all the signals are ok, I have a DIY oscilloscope and I can see the PWM, ZC signals working great and the out AC is chopped as expected.
However, when I connect an oil heater, the AC output is the same as the input, so I don’t know why when no load AC output is correct, but with load AC output is the same as input.
I tried with leading and trailing and I have the same behaviour. 
I can make photos if you want.

glmnet commented:
> I’ve problems with the dimmer, I have a robotdyn with ESP32 and all the signals are ok, I have a DIY oscilloscope and I can see the PWM, ZC signals working great and the out AC is chopped as expected.
> 
> However, when I connect an oil heater, the AC output is the same as the input, so I don’t know why when no load AC output is correct, but with load AC output is the same as input.
> 
> I tried with leading and trailing and I have the same behaviour. 
> 
> I can make photos if you want.
> 
> 

What’s the triac part number?

djares01 commented:
> What’s the triac part number?
> 
> Also. Do you have other loads to try?

Its a BTA16 600B, I've tried with a 50W halogen bulb and I see the same, no difference in light. Maybe robotdyn is broken?

glmnet commented:
I don't think so, however I really don't know what might be wrong. In leading pulse mode, the output is just a pulse, I'm just a SW developer and don't know electronics deep enough to diagnose that issue, I believe it should be easier for someone with a better electrics background, you might post this issue in some electrical engineering forum like the https://www.eevblog.com/ or the https://electronics.stackexchange.com/ I don't because I don't have failing hardware where I can take precise measurements

HFsi commented:
> > What’s the triac part number?
> > Also. Do you have other loads to try?
> 
> Its a BTA16 600B, I've tried with a 50W halogen bulb and I see the same, no difference in light. Maybe robotdyn is broken?

I've to resolder one of the ICs once, it only turned on and off, after the fix it worked fine

misudo commented:
The custom component directory at the top of this issue takes me to 404. I've got the module and could test it with - what's the best way to get hold of the ac_dimmer for my esphome? Thanks.

glmnet commented:
> The custom component directory at the top of this issue takes me to 404. I've got the module and could test it with - what's the best way to get hold of the ac_dimmer for my esphome? Thanks. 

This PR is already merged. That means this component is available in the dev version. Also docs are available in next.esphome.io

djares01 commented:
> 
> 
> I don't think so, however I really don't know what might be wrong. In leading pulse mode, the output is just a pulse, I'm just a SW developer and don't know electronics deep enough to diagnose that issue, I believe it should be easier for someone with a better electrics background, you might post this issue in some electrical engineering forum like the https://www.eevblog.com/ or the https://electronics.stackexchange.com/ I don't because I don't have _failing hardware_ where I can take precise measurements

I've tried the code from this example from robotdyn and for me it works https://github.com/RobotDynOfficial/RBDDimmer/tree/master/examples/SerialMonitorDim

This is my oscilloscope at 50%, I can't show the full signal because my oscilloscope can only show 300V and the AC signals goes from -330V to 330V, its leading pulse.

![IMG_6245](https://user-images.githubusercontent.com/58332149/79594358-27ad0680-80dd-11ea-91b3-4f018d17398a.jpg)

I've tried to capture the PWM and ZC signal with this code, but I don't why I can't, maybe it's generating a different signal than ac-dimmer component.

Thanks for your help!
